apache/trafficcontrol
# EDNS0 Client Subnet for Delivery Services

## What does this PR (Pull Request) do?

This PR adds the ability to enable EDNS0 client subnet on the delivery service level.  Currently, this can be enabled on the CDN level, and this PR still supports that functionality but allows for enabling it per delivery service as well.

- [x] This PR fixes #2929


## Which Traffic Control components are affected by this PR?


- Documentation
- Traffic Ops
- Traffic Portal
- Traffic Router

## What is the best way to verify this PR?


- Verify that the Delivery Services page in Traffic Portal has been updated in the Routing section to include a new boolean value for EcsEnabled.
- Verify that updating this value gets updated in the TO database.
- Verify that the API delivery services endpoints work as expected.
- In a public facing environment, update a DNS Delivery Service to be Ecs Enabled. Perform a snapshot.
- Run the following commands replacing yourDeliverySevice with your delivery service edge location (ex: edge.example.com), yourSubnet and yourSubnet2 with a valid subnets from 2 far apart locations (can be found from /opt/traffic_router/db/czmap.json):
        dig +trace @8.8.8.8 yourDeliverySevice +dnssec ALL
        dig +trace @8.8.8.8 yourDeliverySevice +dnssec +subnet=yourSubnet ALL
        dig +trace @8.8.8.8 yourDeliverySevice +dnssec +subnet=yourSubnet2 ALL
- In Traffic Router (could be in any of the active Traffic Routers) view the access log and find the entry for your requests.  
- Verify that the first request has "chi=anIPAddress rhi=-" and that ans= list of IPs of the edge cache/s closest to YOU.
- Verify that the second request has "chi=anIPAddressInYourSubnet rhi=IPAddressFromFirstRequest" and that ans = list of IPs of the edge cache/s closest to the subnet location.
- Verify that the third request has "chi=anIPAddressInYourSubnet2 rhi=IPAddressFromFirstRequest" and that ans = list of IPs of the edge cache/s closest to the subnet2 location.  

- Follow steps again for a different DNS Delivery Service but leave Ecs Disabled and verify that all requests get "chi=anIPAddress rhi=-" and that ans= list of IPs of the edge cache/s closest to YOU.

## The following criteria are ALL met by this PR




## Additional Information
